spec/sql_database_spec.rb

```diff
@@ -0,0 +1,98 @@
+# frozen_string_literal: true
+
+require 'sql_database'
+
+RSpec.describe SQLDatabase do
+  let(:database) { described_class.new('file::memory:?cache=shared') }
+  let(:sql) { SQLite3::Database.open('file::memory:?cache=shared') }
+  let(:weather_data) { JSON.load_file('fixtures/london_weather.json', { symbolize_names: true }) }
+
+  before { database }
+  after { sql.execute 'DROP TABLE weather' }
+
+  describe '#augment_database' do
+    it 'adds weather information to the database' do
+      database.augment_database(weather_data)
+
+      result = sql.query "SELECT * FROM weather WHERE name = 'London'" do |rows|
+        rows.next_hash.transform_keys(&:to_sym)
+      end
+
+      expect(result).to eq(london_weather_output)
+    end
+
+    it 'updates the database if the same weather information already exists' do
+      database.augment_database(weather_data)
+      database.augment_database(weather_data)
+
+      result = sql.query "SELECT * FROM weather WHERE name = 'London'"
+
+      expect(result.count).to eq(1)
+    end
+
+    it 'returns a success message if no expections are made' do
+      result = database.augment_database(weather_data)
+
+      expect(result).to include('Successful')
+    end
```

Comment:
I think I've figured out what's the thing you are confused about.

You are confusing **error handling** with **return values**. Those two things should be handled separately.

Checking for string values like this is a code smell.
There is no need to return a success message here.
That would be something you could decide to do in your user interface if everything went well. 
"_Success message_" and "_Error message_" are concepts managed by a UI, not a database.

The business logic does not usually return messages for humans (there are exceptions like everything in life, but it's not usual).

The return value here would be whatever `db.execute` returns. 
That's your **happy path**.
If everything goes right, SQLite's `execute` method returns an empty array, so you could have:

```ruby
expect(result).to be_empty
```
or you could decide to intentionally return `true` (for example, that's what Rails returns).

To handle the errors, you could either rescue the exception in the caller of `database.augment`, or if you want to rescue here, then you can flag that there was an error somehow. But **you need to have a return value too**.

Your return value from the error path can be anything that `execute` would return, but that you can not use, like a null thing. 
This is were the **Null pattern** comes in handy. You could send something that behaves like the happy path, but does nothing, or explodes back, or is unusable, etc. etc., depending on how you are going to use it (for example, `retrieve_weather` returns a hash with things, but the unhappy path returns an empty hash, you can't do nothing with it, but is also a hash).

You could also just return `false` if you decide to return `true` in the happy path.
You could set an error message somewhere to use it later, or just set a boolean flag, or append an error code to your application's error handler, if you had one.

There are many ways you can handle an error. 
But error handling and return value are two different things.
You can't build an application based on strings returned from here and there.